crates/red_knot_python_semantic/src/types/narrow.rs

```diff
@@ -194,6 +201,52 @@ impl<'db> NarrowingConstraintsBuilder<'db> {
         }
     }
 
+    /// Generate a constraint from the *type* of the second argument of an `isinstance` call.
+    ///
+    /// Example: for `isinstance(…, str)`, we would infer `Type::ClassLiteral(str)` from the
+    /// second argument, but we need to generate a `Type::Instance(str)` constraint that can
+    /// be used to narrow down the type of the first argument.
+    fn to_isinstance_constraint(&self, classinfo: &Type<'db>) -> Option<Type<'db>> {
+        match classinfo {
+            Type::ClassLiteral(class) => Some(Type::Instance(*class)),
+            Type::Tuple(tuple) => {
+                let mut builder = UnionBuilder::new(self.db);
+                for element in tuple.elements(self.db) {
+                    builder = builder.add(self.to_isinstance_constraint(element)?);
+                }
+                Some(builder.build())
+            }
+            _ => None,
+        }
+    }
+
+    fn add_expr_call(&mut self, expr_call: &ast::ExprCall, expression: Expression<'db>) {
+        let scope = self.scope();
+        let inference = infer_expression_types(self.db, expression);
+
+        if let Some(func_type) = inference
+            .expression_ty(expr_call.func.scoped_ast_id(self.db, scope))
+            .into_function_literal_type()
+        {
+            if func_type.known(self.db) == Some(KnownFunction::IsInstance)
```

Comment:
nit: let's add a `FunctionType::is_known` method, for symmetry with `ClassType::is_known`:

https://github.com/astral-sh/ruff/blob/72c18c8225166f1f08a826c622b9cbcb48331162/crates/red_knot_python_semantic/src/types.rs#L1622-L1627